File: frontend/src/api/report.ts
import client from './client';

export interface Report {
  id: number;
  reporterId: number;
  targetType: 'POST' | 'COMMENT';
  targetId: number;
  reason: string;
  status: 'PENDING' | 'REVIEWED' | 'RESOLVED' | 'DISMISSED';
  createdAt: string;
}

export interface SubmitReportRequest {
  contentType: 'POST' | 'COMMENT';
  contentId: number;
  reason: string;
  description?: string;
}

/**
 * Submit a report
 * POST /api/reports
 */
export async function submitReport(data: SubmitReportRequest): Promise<Report> {
  const response = await client.post<{ success: boolean; message: string; data: Report }>('/reports', data);
  return response.data.data;
}

/**
 * Get all reports (admin)
 * GET /api/admin/reports
 */
export async function adminGetReports(): Promise<Report[]> {
  const response = await client.get<Report[]>('/admin/reports');
  return response.data;
}

/**
 * Handle a report status
 * PUT /api/admin/reports/{id}
 */
export async function adminHandleReport(id: number, status: 'RESOLVED' | 'DISMISSED'): Promise<void> {
  await client.put(`/admin/reports/${id}`, { status });
}

File: microservices/content-service/src/main/java/com/cc91/contentservice/entity/Report.java
package com.cc91.contentservice.entity;

import jakarta.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "reports", indexes = {
    @Index(name = "idx_reports_status", columnList = "status"),
    @Index(name = "idx_reports_target", columnList = "target_type, target_id")
})
public class Report {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "reporter_id", nullable = false)
    private Long reporterId;

    @Enumerated(EnumType.STRING)
    @Column(name = "target_type", nullable = false, length = 10)
    private TargetType targetType;

    @Column(name = "target_id", nullable = false)
    private Long targetId;

    @Column(nullable = false, length = 500)
    private String reason;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 10)
    private ReportStatus status = ReportStatus.PENDING;

    @Column(name = "admin_comment", length = 500)
    private String adminComment;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "reviewed_at")
    private LocalDateTime reviewedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }

    public Report() {}

    public Report(Long reporterId, TargetType targetType, Long targetId, String reason) {
        this.reporterId = reporterId;
        this.targetType = targetType;
        this.targetId = targetId;
        this.reason = reason;
    }

    public enum TargetType { POST, COMMENT }
    public enum ReportStatus { PENDING, REVIEWED, RESOLVED, DISMISSED }

    // Getters and Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public Long getReporterId() { return reporterId; }
    public void setReporterId(Long reporterId) { this.reporterId = reporterId; }

    public TargetType getTargetType() { return targetType; }
    public void setTargetType(TargetType targetType) { this.targetType = targetType; }

    public Long getTargetId() { return targetId; }
    public void setTargetId(Long targetId) { this.targetId = targetId; }

    public String getReason() { return reason; }
    public void setReason(String reason) { this.reason = reason; }

    public ReportStatus getStatus() { return status; }
    public void setStatus(ReportStatus status) { this.status = status; }

    public String getAdminComment() { return adminComment; }
    public void setAdminComment(String adminComment) { this.adminComment = adminComment; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

    public LocalDateTime getReviewedAt() { return reviewedAt; }
    public void setReviewedAt(LocalDateTime reviewedAt) { this.reviewedAt = reviewedAt; }
}
